src/test/java/com/ibm/cloud/sdk/core/test/security/Cp4dServiceInstanceAuthenticatorTest.java

```diff
@@ -0,0 +1,507 @@
+    assertNotNull(tokenServerRequest);
+    assertNotNull(tokenServerRequest.getHeaders());
+    Headers actualHeaders = tokenServerRequest.getHeaders();
+    assertEquals("value1", actualHeaders.get("header1"));
+    assertEquals("value2", actualHeaders.get("header2"));
+
+    // Authenticator should just return the same token this time since we have a valid one stored.
+    requestBuilder = new Request.Builder().url("https://test.com");
+    authenticator.authenticate(requestBuilder);
+    verifyAuthHeader(requestBuilder, "Bearer " + tokenData.getData().getToken());
+  }
+
+  @Test
+  public void testApiErrorBadRequest() throws Throwable {
+    server.enqueue(errorResponse(400));
+
+    // Mock current time to ensure the token is valid.
+    PowerMockito.mockStatic(Clock.class);
+    PowerMockito.when(Clock.getCurrentTimeInSeconds()).thenReturn((long) 100);
+
+    CloudPakForDataServiceInstanceAuthenticator authenticator = new CloudPakForDataServiceInstanceAuthenticator.Builder()
+        .url(url)
+        .username(testUsername)
+        .apikey(testApikey)
+        .serviceInstanceId(testServiceInstanceId)
+        .build();
+
+    Request.Builder requestBuilder = new Request.Builder().url("https://test.com");
+
+    // Calling authenticate should result in an exception.
+    try {
+      authenticator.authenticate(requestBuilder);
+      fail("Expected authenticate() to result in exception!");
+    } catch (ServiceResponseException excp) {
+      assertTrue(excp instanceof ServiceResponseException);
+    } catch (Throwable t) {
+      fail("Expected ServiceResponseException, not " + t.getClass().getSimpleName());
+    }
+  }
+
+  @Test
+  public void testApiErrorResponse() throws Throwable {
+    server.enqueue(jsonResponse("{'}"));
+
+    // Mock current time to ensure the token is valid.
+    PowerMockito.mockStatic(Clock.class);
+    PowerMockito.when(Clock.getCurrentTimeInSeconds()).thenReturn((long) 100);
+
+    CloudPakForDataServiceInstanceAuthenticator authenticator = new CloudPakForDataServiceInstanceAuthenticator.Builder()
+        .url(url)
+        .username(testUsername)
+        .apikey(testApikey)
+        .serviceInstanceId(testServiceInstanceId)
+        .build();
+
+    Request.Builder requestBuilder = new Request.Builder().url("https://test.com");
+
+    // Calling authenticate should result in an exception.
+    try {
+      authenticator.authenticate(requestBuilder);
+      fail("Expected authenticate() to result in exception!");
+    } catch (RuntimeException excp) {
+      Throwable causedBy = excp.getCause();
+      assertNotNull(causedBy);
+      assertTrue(causedBy instanceof IllegalStateException);
+    } catch (Throwable t) {
+      fail("Expected RuntimeException, not " + t.getClass().getSimpleName());
+    }
+  }
+
+  @Test(expectedExceptions =IllegalArgumentException.class)
+  public void testCtorMissingUsernameMap() {
+    Map<String, String> props = new HashMap<>();
```

Comment:
it looks like this test and the rest of the tests in this class are duplicates of testcases in the first section of test methods.